core/src/main/scala/cilib/Archive.scala

```diff
@@ -0,0 +1,195 @@
+      case Empty(b) => NonEmpty[A, B](==>>.singleton(k, v), b)
+      case NonEmpty(m, b) =>
+        b match {
+          case Bounded(limit) =>
+            if (limit.value <= m.size) NonEmpty[A, B](m.insert(k, v), b)
+            else NonEmpty[A, B](m, b)
+          case Unbounded() => NonEmpty[A, B](m, b)
+        }
+    }
+
+  def insertWithCondition(f: (B, B) => Boolean)(k: A, v: B): Archive[A, B] =
+    this match {
+      case Empty(b) => NonEmpty[A, B](==>>.singleton(k, v), b)
+      case NonEmpty(m, b) =>
+        b match {
+          case Bounded(limit) =>
+            if (limit.value <= m.size && m.values.forall(x => f(v, x)))
+              NonEmpty[A, B](m.insert(k, v), b)
+            else NonEmpty[A, B](m, b)
+          case Unbounded() => NonEmpty[A, B](m, b)
+        }
+
+    }
+
+  def deleteByKey(k: A): Archive[A, B] =
+    this match {
+      case Empty(b) => Empty[A, B](b)
+      case NonEmpty(m, b) =>
+        val newMap = m.delete(k)
+        if (newMap.isEmpty) Empty[A, B](b)
+        else NonEmpty(newMap, b)
+    }
+
+  def deleteByValue(v: B): Archive[A, B] =
+    this match {
+      case Empty(b) => Empty[A, B](b)
+      case NonEmpty(m, b) =>
+        val newMap = m.filter(x => x != v)
+        if (newMap.isEmpty) Empty[A, B](b)
+        else NonEmpty(newMap, b)
+    }
+
+  def deleteWithCondition(f: B => Boolean): Archive[A, B] =
+    this match {
+      case Empty(b) => Empty(b)
+      case NonEmpty(m, b) =>
+        val newMap = m.filter(x => f(x))
+        if (newMap.isEmpty) Empty[A, B](b)
+        else NonEmpty(newMap, b)
+    }
+
+  def deleteMax: Archive[A, B] =
+    this match {
+      case Empty(b)       => Empty[A, B](b)
+      case NonEmpty(m, b) => NonEmpty(m.deleteMax, b)
+    }
+
+  def deleteMin: Archive[A, B] =
+    this match {
+      case Empty(b)       => Empty[A, B](b)
+      case NonEmpty(m, b) => NonEmpty(m.deleteMin, b)
+    }
+
+  def findMax: Option[(A, B)] =
+    this match {
+      case Empty(_)       => None
+      case NonEmpty(m, _) => m.findMax
+    }
+
+  def findMin: Option[(A, B)] =
+    this match {
+      case Empty(_)       => None
+      case NonEmpty(m, _) => m.findMin
+    }
+
+  def adjust(k: A, f: B => B): Archive[A, B] =
+    this match {
+      case Empty(b)       => Empty(b)
+      case NonEmpty(m, b) => NonEmpty(m.adjust(k, f), b)
+    }
+
+  def adjustWithKey(k: A, f: (A, B) => B): Archive[A, B] =
+    this match {
+      case Empty(b)       => Empty(b)
+      case NonEmpty(m, b) => NonEmpty(m.adjustWithKey(k, f), b)
+    }
+
+  def dominates(f: (B, B) => Boolean)(v: B): Boolean =
+    this match {
+      case Empty(_)       => true
+      case NonEmpty(m, _) => m.values.forall(x => f(v, x))
+    }
+
+  def elemAt(i: Int): Option[(A, B)] =
+    this match {
+      case Empty(_)       => None
+      case NonEmpty(m, _) => m.elemAt(i)
+    }
+
+  def empty: Empty[A, B] =
+    this match {
+      case Empty(b)       => Empty(b)
+      case NonEmpty(_, b) => Empty(b)
```

Comment:
Why even pattern match? The data structure is just being replaced so you can reuse the bounds function and then return an `Empty`: `Empty(this.bound)`.

Do you have specific cases for this function?

Reply:
I don't have a specific case. My thinking was if a user needed to delete everything from the archive it  would be easier to create a `Empty`.